include/utap/string_interning.h

```diff
@@ -1,51 +1,67 @@
-#pragma once
-#include <algorithm>
-#include <list>
-#include <string>
+/* libutap - Uppaal Timed Automata Parser.
+   Copyright (C) 2020 Aalborg University.
+   Copyright (C) 2002-2006 Uppsala University and Aalborg University.
 
-namespace UTAP {
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public License
+   as published by the Free Software Foundation; either version 2.1 of
+   the License, or (at your option) any later version.
 
-/**
- * Special type used distinquish a int from a string index, also stores a reference to string for ease of use
- */
-class StringIndex
-{
-    friend class InternedStringVector;
+   This library is distributed in the hope that it will be useful, but
+   WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
 
-    std::string_view string;
-    int interned_index;
+   You should have received a copy of the GNU Lesser General Public
+   License along with this library; if not, write to the Free Software
+   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
+   USA
+*/
 
-    StringIndex(std::string_view str, int index): string{str}, interned_index{index} {}
+#ifndef UTAP_STRINGINTERNING_H
+#define UTAP_STRINGINTERNING_H
 
-public:
-    bool operator==(const StringIndex& other) const { return interned_index == other.interned_index; }
+#include <algorithm>
+#include <string>
+#include <vector>
 
-    std::string_view str() const { return string; }
-    int index() const { return interned_index; }
-};
+namespace UTAP {
 
 /**
- * Intern strings to reduce memory footprints
+ * Reuses strings to reduce memory footprint
  */
-class InternedStringVector
+class InternedStrings
 {
 public:
-    const std::list<std::string>& get_strings() const { return strings; }
-    StringIndex add_string_if_new(std::string&& string)
+    class Index
+    {
+        const InternedStrings* interned;  // pointer allows default copy operations
+        size_t id;
+
+    public:
+        Index(const InternedStrings& interned, size_t id): interned{&interned}, id{id} {}
+        size_t index() const { return id; }
+        const std::string& str() const { return interned->strings[id]; }
+        bool operator==(const Index& other) const { return interned == other.interned && id == other.id; }
```

Comment:
I don't think we need a custom equality operator here. I think you can just delete this

Reply:
Well, I implemented because previous implementation had it, and it is actually used.
I think compiler cannot supply equality operator default due to private members.

Reply:
Ahh you're right, and we can't even `default` the implementation because that's a c++20 feature...